lib/DBus/Generation.hs

```diff
@@ -0,0 +1,539 @@
+import           Data.Maybe
+import           Data.Monoid
+import           Data.String
+import           Data.Traversable
+import           Data.Word
+import           Language.Haskell.TH
+import           Prelude hiding (mapM)
+import           System.IO.Unsafe
+import           System.Posix.Types (Fd(..))
+
+type ClientBusPathR a = ReaderT (Client, T.BusName, T.ObjectPath) IO a
+
+dbusInvoke :: (Client -> T.BusName -> T.ObjectPath -> a) -> ClientBusPathR a
+dbusInvoke fn = (\(c, b, p) -> fn c b p) <$> ask
+
+-- Use these operators together with dbusInvoke to invoke functions of the form
+-- Client -> T.BusName -> T.ObjectPath
+infixl 4 ??
+(??) :: Functor f => f (a -> b) -> a -> f b
+fab ?? a = fmap ($ a) fab
+{-# INLINE (??) #-}
+
+infixl 4 ?/?
+(?/?) :: ClientBusPathR (a -> IO b) -> a -> ClientBusPathR b
+soFar ?/? arg = do
+  returnValue <- fmap ($ arg) soFar
+  lift returnValue
+
+data GenerationParams = GenerationParams
+  { genBusName :: Maybe T.BusName
+  , genObjectPath :: Maybe T.ObjectPath
+  , genInterfaceName :: T.InterfaceName
+  , genTakeSignalErrorHandler :: Bool
+  , getTHType :: T.Type -> Type
+  , useDBusR :: Bool
+  }
+
+defaultGetDictType :: Type -> Type -> Type
+defaultGetDictType k v =
+  AppT (AppT (ConT ''Map.Map) k) (v)
+
+defaultGetTHType :: T.Type -> Type
+defaultGetTHType = buildGetTHType (AppT ListT) defaultGetDictType
+
+buildGetTHType ::
+  (Type -> Type) -> (Type -> Type -> Type) -> T.Type -> Type
+buildGetTHType arrayTypeBuilder dictTypeBuilder = fn
+  where fn t =
+          case t of
+            -- Because of a quirk in how we unmarshal things, we currently HAVE
+            -- to decorde arrays of Word8 in this way.
+            T.TypeArray T.TypeWord8 -> ConT ''BS.ByteString
+            T.TypeBoolean -> ConT ''Bool
+            T.TypeWord8 -> ConT ''Word8
+            T.TypeWord16 -> ConT ''Word16
+            T.TypeWord32 -> ConT ''Word32
+            T.TypeWord64 -> ConT ''Word64
+            T.TypeInt16 -> ConT ''Int16
+            T.TypeInt32 -> ConT ''Int32
+            T.TypeInt64 -> ConT ''Int64
+            T.TypeDouble -> ConT ''Double
+            T.TypeUnixFd -> ConT ''Fd
+            T.TypeString -> ConT ''String
+            T.TypeSignature -> ConT ''T.Signature
+            T.TypeObjectPath -> ConT ''T.ObjectPath
+            T.TypeVariant -> ConT ''T.Variant
+            T.TypeArray arrayType -> arrayTypeBuilder $ fn arrayType
+            T.TypeDictionary k v -> dictTypeBuilder (fn k) (fn v)
+            T.TypeStructure ts -> foldl AppT (TupleT $ length ts) $ map fn ts
+
+newNameDef :: [Char] -> Q Name
+newNameDef n =
+  case n of
+    "" -> newName "arg"
+    "data" -> newName "arg"
+    _ -> newName n
+
+defaultGenerationParams :: GenerationParams
+defaultGenerationParams =
+  GenerationParams
+  { genBusName = Nothing
+  , genInterfaceName = fromString ""
+  , getTHType = defaultGetTHType
+  , genObjectPath = Nothing
+  , genTakeSignalErrorHandler = False
```

Comment:
/home/travis/build/rblaze/haskell-dbus/lib/DBus/Generation.hs:96:3: warning: [-Wmissing-fields]
    • Fields of ‘GenerationParams’ not initialised: useDBusR
    • In the expression:
        GenerationParams
          {genBusName = Nothing, genInterfaceName = fromString "",
           getTHType = defaultGetTHType, genObjectPath = Nothing,
           genTakeSignalErrorHandler = False}